pkg/minikube/node/config.go

```diff
@@ -127,12 +129,24 @@ func setupKubeAdm(mAPI libmachine.API, cfg config.ClusterConfig, node config.Nod
 }
 
 func setupKubeconfig(h *host.Host, c *config.ClusterConfig, n *config.Node, clusterName string) (*kubeconfig.Settings, error) {
-	addr, err := h.Driver.GetURL()
-	if err != nil {
-		exit.WithError("Failed to get driver URL", err)
-	}
-	if !driver.IsKIC(h.DriverName) {
+	addr := ""
+	var err error
+	if driver.IsKIC(h.DriverName) {
+		p, err := oci.HostPortBinding(h.DriverName, h.Name, n.Port)
+		if err != nil {
+			exit.WithError("Failed to get host binding port for api port", err)
+		}
+		addr = fmt.Sprintf("https://" + net.JoinHostPort("127.0.0.1", strconv.Itoa(p)))
+	} else {
+		addr, err = h.Driver.GetURL()
+		if err != nil {
+			exit.WithError("Failed to get driver URL", err)
+		}
+
 		addr = strings.Replace(addr, "tcp://", "https://", -1)
+		// because of libmachine problem with detecting docker port https://github.com/kubernetes/minikube/issues/7017
+		// we have to do this hack https://github.com/docker/machine/blob/master/libmachine/provision/utils.go#L159_L175
+		// contact @afbjorklund for more details
 		addr = strings.Replace(addr, ":2376", ":"+strconv.Itoa(n.Port), -1)
```

Comment:
This code is getting hard to follow, and I especially think it would be cleaner to call GetIP instead of search and replacing hardcoded numbers. What do you think about extracting this logic into a function like:

```
func apiServerURL() {
  if KIC() {
     p = oci.HostPortBinding(h.DriverName, h.Name, n.Port)
     return fmt.Sprintf("https://" + net.JoinHostPort(oci.DefaultBindIPV4, strconv.Itoa(p))) 
  }
  ip := h.Driver.GetIP()
  return fmt.Sprintf("https://" + net.JoinHostPort(ip, strconv.Itoa(p)))
}
```

Reply:
good suggestion. done